tests/graph_components/providers/test_domain_for_core_training_provider.py

```diff
@@ -0,0 +1,163 @@
+
+        # replaced with default values
+        if key in ["config", SESSION_CONFIG_KEY]:
+            assert modified_dict[key] == default_dict[key]
+
+        # for slots, we drop the specification of how they are filled
+        elif key == KEY_SLOTS:
+            for slot_key in original_dict[key]:
+                assert modified_dict[key][slot_key][SLOT_MAPPINGS] == []
+                original_dict[key][slot_key][SLOT_MAPPINGS] = []  # drop for comparison
+                assert original_dict[key][slot_key] == modified_dict[key][slot_key]
+
+        # for responses, we only keep the keys
+        elif key == KEY_RESPONSES:
+            assert set(modified_dict[key].keys()) == set(original_dict[key].keys())
+            for sub_key in original_dict[key]:
+                assert modified_dict[key][sub_key] == []
+
+        # for forms, we only keep the keys (and the Domain will add a default key)
+        elif key == KEY_FORMS:
+            assert set(modified_dict[key].keys()) == set(original_dict[key].keys())
+            for sub_key in original_dict[key]:
+                assert set(modified_dict[key][sub_key].keys()) == {REQUIRED_SLOTS_KEY}
+                assert modified_dict[key][sub_key][REQUIRED_SLOTS_KEY] == []
+
+        # everything else remains unchanged
+        else:
+            assert original_dict[key] == modified_dict[key]
+
+
+def test_train_core_with_original_or_provided_domain_and_compare(
+    tmp_path_factory: TempPathFactory,
+    default_model_storage: ModelStorage,
+    default_execution_context: ExecutionContext,
+):
+    # Choose an example where the provider will remove a lot of information:
+    example = Path("examples/formbot/")
+    training_files = [example / "data" / "rules.yml"]
+
+    # Choose a configuration with a policy
+    # Note: This is sufficient to illustrate that the component won't be re-trained
+    # when the domain changes. We do *not* test here whether removing keys would/
+    # should not have any effect.
+    config = """
+    recipe: default.v1
+    language: en
+
+    policies:
+      - name: RulePolicy
+    """
+    config_dir = tmp_path_factory.mktemp("config dir")
+    config_file = config_dir / "config.yml"
+    with open(config_file, "w") as f:
+        f.write(config)
+
+    # Train with the original domain
+    original_domain_file = example / "domain.yml"
+    original_output_dir = tmp_path_factory.mktemp("output dir")
+    model_training.train(
+        domain=original_domain_file,
+        config=str(config_file),
+        training_files=training_files,
+        output=original_output_dir,
+    )
+
+    # Let the provider create a modified domain
+    original_domain = Domain.from_file(original_domain_file)
+    component = DomainForCoreTrainingProvider.create(
+        {"arbitrary-unused": 234},
+        default_model_storage,
+        Resource("xy"),
+        default_execution_context,
+    )
+    modified_domain = component.provide(domain=original_domain)
```

Comment:
Isn't this already going to happen as it's part of the default recipe? Or am I missing something?

Reply:
If I don't modify the domain, then this test is pointless, so...
>  I'm basically using the provider to modify the domain instead of removing keys from the domain manually -- and then pass in that modified domain for re-training and check that the policy would be loaded from cache. [...] I could add the same changes to the domain manually again, but the other test should already take care of asserting that the respective keys are really removed/replaced, no? 🤔